fpga_interchange/cost_map.cc

```diff
@@ -0,0 +1,356 @@
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "cost_map.h"
+
+#include "context.h"
+#include "log.h"
+
+NEXTPNR_NAMESPACE_BEGIN
+
+///@brief Factor to adjust the penalty calculation for deltas outside the segment bounding box:
+//      factor < 1.0: penalty has less impact on the final returned delay
+//      factor > 1.0: penalty has more impact on the final returned delay
+static constexpr float PENALTY_FACTOR = 1.f;
+
+///@brief Minimum penalty cost that is added when penalizing a delta outside the segment bounding box.
+static constexpr delay_t PENALTY_MIN = 1;
+
+// also known as the L1 norm
+static int manhattan_distance(const std::pair<int32_t, int32_t> &a, const std::pair<int32_t, int32_t> &b)
+{
+    return std::abs(b.first - a.first) + std::abs(b.second - a.second);
+}
+
+static delay_t penalize(const delay_t &entry, int distance, delay_t penalty)
+{
+    penalty = std::max(penalty, PENALTY_MIN);
+    return entry + distance * penalty * PENALTY_FACTOR;
+}
+
+delay_t CostMap::get_delay(const Context *ctx, WireId src_wire, WireId dst_wire) const
+{
+    TypeWirePair type_pair;
+    type_pair.src = TypeWireId(ctx, src_wire);
+    type_pair.dst = TypeWireId(ctx, dst_wire);
+
+    int src_tile;
+    if (src_wire.tile == -1) {
+        src_tile = ctx->chip_info->nodes[src_wire.index].tile_wires[0].tile;
+    } else {
+        src_tile = src_wire.tile;
+    }
+
+    int32_t src_x, src_y;
+    ctx->get_tile_x_y(src_tile, &src_x, &src_y);
+
+    int dst_tile;
+    if (dst_wire.tile == -1) {
+        dst_tile = ctx->chip_info->nodes[dst_wire.index].tile_wires[0].tile;
+    } else {
+        dst_tile = dst_wire.tile;
+    }
+
+    int32_t dst_x, dst_y;
+    ctx->get_tile_x_y(dst_tile, &dst_x, &dst_y);
+
+    auto iter = cost_map_.find(type_pair);
+    if (iter == cost_map_.end()) {
+        auto &src_type = ctx->chip_info->tile_types[type_pair.src.type];
+        IdString src_tile_type(src_type.name);
+        IdString src_wire_name(src_type.wire_data[type_pair.src.index].name);
+
+        auto &dst_type = ctx->chip_info->tile_types[type_pair.dst.type];
+        IdString dst_tile_type(dst_type.name);
+        IdString dst_wire_name(dst_type.wire_data[type_pair.dst.index].name);
+
+#if 0
+        log_warning("Delay matrix is missing %s/%s -> %s/%s\n",
+                src_tile_type.c_str(ctx),
+                src_wire_name.c_str(ctx),
+                dst_tile_type.c_str(ctx),
+                dst_wire_name.c_str(ctx));
+#endif
+        return std::numeric_limits<delay_t>::max();
+    }
+
+    const auto &delay_matrix = iter->second;
+
+    int32_t off_x = delay_matrix.offset.first + (dst_x - src_x);
+    int32_t off_y = delay_matrix.offset.second + (dst_y - src_y);
+
+    int32_t closest_x = off_x;
+    int32_t closest_y = off_y;
+    if (closest_x < 0) {
```

Comment:
using `std::max` and `std::min` would be neater here

Reply:
Fixed